File: phone/app/src/main/java/net/novucs/slither/PlayerConnection.java
package net.novucs.slither;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothProfile;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.atomic.AtomicBoolean;

public class PlayerConnection extends BluetoothGattCallback {

    private final AtomicBoolean servicesFound = new AtomicBoolean(false);
    private final LinkedBlockingQueue<SubscriptionRequest> subscriptionRequests = new LinkedBlockingQueue<>();
    private final LinkedList<SubscriptionRequest> bleSafeRequests = new LinkedList<>();
    private final Map<Subscription, OnNotification> subscriptions = new HashMap<>();
    private final MenuActivity activity;
    private final BluetoothDevice device;
    private Runnable disconnectCallback;
    private BluetoothGatt gatt;
    private boolean connected = false;

    public PlayerConnection(MenuActivity activity, BluetoothDevice device) {
        this.activity = activity;
        this.device = device;
    }

    public BluetoothGatt getGatt() {
        return gatt;
    }

    public void connect(Runnable disconnectCallback) {
        if (connected) {
            throw new IllegalStateException("Already connected");
        }

        this.disconnectCallback = disconnectCallback;
        connected = true;
        gatt = device.connectGatt(activity, false, this);
    }

    @Override
    public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
        if (newState == BluetoothProfile.STATE_CONNECTED) {
            gatt.discoverServices();
        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
            disconnectCallback.run();
        }
    }

    @Override
    public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
        Subscription subscription = Subscription.of(characteristic);
        OnNotification onNotification = subscriptions.get(subscription);
        if (onNotification != null) {
            onNotification.onNotification(characteristic.getValue());
        }
    }

    @Override
    public void onServicesDiscovered(BluetoothGatt gatt, int status) {
        servicesFound.set(true);
        subscriptionRequests.drainTo(bleSafeRequests);

        if (bleSafeRequests.isEmpty()) {
            return;
        }

        SubscriptionRequest request = bleSafeRequests.pop();

        if (request != null) {
            subscribe(request.getCallback(), request.getServiceId(), request.getCharacteristicId());
        }
    }

    @Override
    public void onDescriptorWrite(BluetoothGatt gatt, BluetoothGattDescriptor descriptor, int status) {
        subscriptionRequests.drainTo(bleSafeRequests);

        if (bleSafeRequests.isEmpty()) {
            return;
        }

        SubscriptionRequest request = bleSafeRequests.pop();

        if (request != null) {
            subscribe(request.getCallback(), request.getServiceId(), request.getCharacteristicId());
        }
    }

    public void subscribe(OnNotification callback, UUID serviceId, UUID characteristicId)
            throws IllegalArgumentException {
        if (!servicesFound.get()) {
            subscriptionRequests.add(new SubscriptionRequest(callback, serviceId, characteristicId));
            return;
        }

        BluetoothGattService service = gatt.getService(serviceId);
        if (service == null) {
            gatt.disconnect();
            return;
        }

        BluetoothGattCharacteristic characteristic = service.getCharacteristic(characteristicId);
        if (characteristic == null) {
            gatt.disconnect();
            return;
        }

        if (characteristic.getDescriptors().size() == 0) {
            gatt.disconnect();
            return;
        }

        gatt.setCharacteristicNotification(characteristic, true);
        BluetoothGattDescriptor descriptor = characteristic.getDescriptors().get(0);
        descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
        gatt.writeDescriptor(descriptor);

        Subscription subscription = Subscription.of(characteristic);
        subscriptions.put(subscription, callback);
    }

    public interface OnNotification {
        void onNotification(byte[] data);
    }

    private static class SubscriptionRequest {
        private final OnNotification callback;
        private final UUID serviceId;
        private final UUID characteristicId;

        public SubscriptionRequest(OnNotification callback, UUID serviceId, UUID characteristicId) {
            this.callback = callback;
            this.serviceId = serviceId;
            this.characteristicId = characteristicId;
        }

        public OnNotification getCallback() {
            return callback;
        }

        public UUID getServiceId() {
            return serviceId;
        }

        public UUID getCharacteristicId() {
            return characteristicId;
        }
    }

    private static class Subscription {
        private final UUID serviceId;
        private final UUID characteristicId;

        public Subscription(UUID serviceId, UUID characteristicId) {
            this.serviceId = serviceId;
            this.characteristicId = characteristicId;
        }

        public static Subscription of(BluetoothGattCharacteristic characteristic) {
            return new Subscription(characteristic.getService().getUuid(),
                    characteristic.getUuid());
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Subscription that = (Subscription) o;

            if (serviceId != null ? !serviceId.equals(that.serviceId) : that.serviceId != null)
                return false;
            return characteristicId != null ? characteristicId.equals(that.characteristicId) : that.characteristicId == null;
        }

        @Override
        public int hashCode() {
            int result = serviceId != null ? serviceId.hashCode() : 0;
            result = 31 * result + (characteristicId != null ? characteristicId.hashCode() : 0);
            return result;
        }
    }
}

File: android/app/src/main/java/net/novucs/slither/PlayerConnection.kt
package net.novucs.slither

import android.bluetooth.*
import net.novucs.slither.BLEAttributes.SNAKE_MESSAGE_SERVICE
import net.novucs.slither.BLEAttributes.SNAKE_TX_CHARACTERISTIC
import java.util.*
import java.util.concurrent.LinkedBlockingQueue

/**
 * Handles all bluetooth gatt connection callbacks for a single player.
 * Provides a layer over the android gatt api to synchronise all requests.
 */
class PlayerConnection(private val context: SlitherActivity,
                       private val device: BluetoothDevice) : BluetoothGattCallback() {

    private val operations = LinkedBlockingQueue<() -> Unit>()
    private val subscriptions = mutableMapOf<Subscription, (data: ByteArray) -> Unit>()
    private var disconnectCallback: (() -> Unit)? = null
    private var gatt: BluetoothGatt? = null
    private var connected = false
    private var busy = false

    /**
     * Connects to the found bluetooth device.
     */
    fun connect() {
        if (connected) {
            throw IllegalStateException("Already connected")
        }

        busy = true
        connected = true
        gatt = device.connectGatt(context, false, this)
    }

    /**
     * Sets a handler for disconnecting.
     *
     * @param callback the callback to invoke when the device disconnects.
     */
    fun onDisconnect(callback: () -> Unit) {
        disconnectCallback = callback
    }

    /**
     * Discovers new services on first connection, calls the disconnect
     * callback on disconnection.
     */
    override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
        if (newState == BluetoothProfile.STATE_CONNECTED) {
            gatt.discoverServices()
        } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
            disconnectCallback?.invoke()
        }
    }

    /**
     * Invokes any subscribed callbacks when notified of a characteristic
     * change.
     */
    override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
        val subscription = subscriptionOf(characteristic)
        val callback = subscriptions[subscription]
        callback?.invoke(characteristic.value)
    }

    /**
     * Executes any queued operations after service request has been fulfilled.
     */
    override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
        executeNextOperation()
    }

    /**
     * Executes any queued operations after service write has been fulfilled.
     */
    override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
        executeNextOperation()
    }

    /**
     * Executes any queued operations after service read has been fulfilled.
     */
    override fun onDescriptorRead(gatt: BluetoothGatt?, descriptor: BluetoothGattDescriptor?, status: Int) {
        executeNextOperation()
    }

    /**
     * Sets the service to not busy, polls and invokes the next operation.
     */
    private fun executeNextOperation() {
        busy = false
        operations.poll()?.invoke()
    }

    /**
     * Subscribes to all notifications for a particular characteristic of a
     * given service.
     *
     * @param callback the callback to be invoked on notification.
     * @param serviceId the service ID to subscribe to.
     * @param characteristicId the characteristic ID to subscribe to.
     */
    fun subscribe(callback: (data: ByteArray) -> Unit, serviceId: UUID, characteristicId: UUID) {
        synchronized(busy) {
            if (busy) {
                operations += { subscribe(callback, serviceId, characteristicId) }
                return
            }
            busy = true
        }

        val (gatt, characteristic) = findCharacteristic(serviceId, characteristicId) ?: return

        if (characteristic.descriptors.isEmpty()) {
            gatt.disconnect()
            return
        }

        gatt.setCharacteristicNotification(characteristic, true)
        val descriptor = characteristic.descriptors.first()
        descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
        gatt.writeDescriptor(descriptor)

        val subscription = subscriptionOf(characteristic)
        subscriptions[subscription] = callback
    }

    /**
     * Sends a text message via gatt to the player. The player's micro:bit
     * should then scroll this message to the LED display once sent.
     */
    fun sendMessage(message: String) {
        synchronized(busy) {
            if (busy) {
                operations += { sendMessage(message) }
                return
            }
            busy = true
        }

        val (gatt, characteristic) = findCharacteristic(
                serviceId = SNAKE_MESSAGE_SERVICE,
                characteristicId = SNAKE_TX_CHARACTERISTIC)
                ?: return

        characteristic.value = message.toByteArray()
        gatt.writeCharacteristic(characteristic)
        busy = false
    }

    /**
     * Finds a characteristic, given its service and characteristic IDs.
     * Disconnects from the device if the characteristic was unable to be found.
     *
     * @param serviceId the service to search for.
     * @param characteristicId the characteristic to search for.
     */
    private fun findCharacteristic(serviceId: UUID, characteristicId: UUID):
            Pair<BluetoothGatt, BluetoothGattCharacteristic>? {
        // Attempt to find and return the characteristic.
        val gatt = this.gatt
        gatt?.getService(serviceId)?.getCharacteristic(characteristicId)?.let {
            return Pair(gatt, it)
        }

        // Disconnect from device on failure, as we are incompatible.
        gatt?.disconnect()
        return null
    }

    /**
     * Represents a notification subscription key, using a composite of the
     * listened service ID and characteristic ID.
     */
    private data class Subscription(private val serviceId: UUID, private val characteristicId: UUID)

    /**
     * Constructs a new subscription using a characteristic.
     *
     * @param characteristic the characteristic to create a subscription of.
     */
    private fun subscriptionOf(characteristic: BluetoothGattCharacteristic): Subscription {
        return Subscription(characteristic.service.uuid, characteristic.uuid)
    }
}
